Font Library: move font uploads to a new tab

madhusudhand opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

As suggested by the design, this change moves the font uploads section to a new tab and provides appropriate error messages when upload fails.

It also removes the use of snackbar.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Open font library modal from global style -> typography.
2. Observe that the uploads section is now moved to a new tab.
3. Try uploading fonts.
4. Upon error, it should show an error in the same tab and should not invoke snackbar.

## Screenshots or screencast <!-- if applicable -->

<img width="849" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/5b06e09b-d7fd-4810-80b5-6a35c05a2f3b">

## Related issues:

Fixes: #54504

madhusudhand committed:
move font uploads to a new tab in the modal

jasmussen commented:
This feels clearer to me, given that if you do not permit access to Google though the "Install Fonts" tab, then we'll want to have a top level alternative, which is upload.

I will note that it would be okay to still have the whole modal be a dropzone, and invoke uploading regardless of which tab you are on. That's not a requirement, this is fine too, just noting. 

I got a curious error, though. I was uploading [this font](https://fonts.google.com/specimen/Inclusive+Sans), and got a _green_ error message simply stating, "There was an error installing the fonts". Screenshot and gif:


<img width="699" alt="error message" src="https://github.com/WordPress/gutenberg/assets/1204802/666637fc-3a21-49af-851f-0f09c4d3a8a3">

![upload](https://github.com/WordPress/gutenberg/assets/1204802/5f28aa71-1b93-4f07-8dd4-0908e2681d84)

* What was the error, do you know?
* If we don't know the error, we should not show any error. If we know the error, we should state exactly what went wrong.
* If it is indeed an error, the notice shouldn't be green. It can be an `error` or `warning`.

Note that the font uploaded successfully, as far as I can tell. Was the notice, given it was green for success, meant to actually say something like "Upload successful"? If yes, then we we should _ideally_ avoid such a notice entirely, and instead immediately upon succesfull upload, redirect to the font page itself, with the boxes checked. I.e. this page:

<img width="1392" alt="success" src="https://github.com/WordPress/gutenberg/assets/1204802/f5c8dfb6-5e6a-4e90-b227-35f724230095">

Nice work, this should be able to land in short order!

madhusudhand committed:
fix invalid success message and revert the dropzone to modal

madhusudhand commented:
@jasmussen Thanks for the feedback.

> I will note that it would be okay to still have the whole modal be a dropzone, and invoke uploading regardless of which tab you are on. That's not a requirement, this is fine too, just noting.

Reverted the dropzone to previous state (whole modal being a dropzone). However, it is not yet supported in any tab page yet due the tab panel structure. I can create a followup ticket to add this behaviour.

> Was the notice, given it was green for success, meant to actually say something like "Upload successful"? If yes, then we we should ideally avoid such a notice entirely, and instead immediately upon succesfull upload, redirect to the font page itself, with the boxes checked.

Yes, It is a success message after successful upload. The message was wrong.
As suggested I have removed the notice upon successful upload.
But, with current state of TabPanel component, it is not possible to programmatically switch tabs. I have created #46704 to add support, but unfortunately the v1 is marked for bug fixes only and [there is v2 version](https://github.com/WordPress/gutenberg/pull/53960) in progress. Until then it is not possible to switch tabs.

Please suggest if we should use different approach in the interim. 

> If we don't know the error, we should not show any error. If we know the error, we should state exactly what went wrong.

With the current state of API, we do not have information about the exact error. I have created [this followup issue](https://github.com/WordPress/gutenberg/issues/54777) to address it and removed the generic error for now.

jasmussen commented:
Sounds good.

> Please suggest if we should use different approach in the interim.

If we cannot yet redirect on successful upload, you should at least get the success notice in the interim. But I hope we can build this in short order, it would be good to have for the feature.

madhusudhand committed:
add success notice for font uploads

madhusudhand commented:
> If we cannot yet redirect on successful upload, you should at least get the success notice in the interim. But I hope we can build this in short order, it would be good to have for the feature.

Introduced the success notice. I have created #54779 to address this in future once the Tabs v2 gets merged.

<img width="575" alt="image" src="https://github.com/WordPress/gutenberg/assets/1935113/a80717f3-aeb9-4582-9012-adf683947dca">

madhusudhand committed:
make supported file formats message dynamic based on allowed extensions

ironprogrammer commented on packages/edit-site/src/components/global-styles/font-library-modal/local-fonts.js:
```diff
@@ -126,33 +135,59 @@ function LocalFonts() {
+					{ sprintf(
+						/* translators: %s: allowed font formats: ex: .ttf, .woff and .woff2 */
+						__(
+							'Uploaded fonts will appear up in your library and can be used in your theme after that. Formats %s are supported.'
```
```suggestion
							'Uploaded fonts appear in your library and can be used in your theme. Supported formats: %s.'
```
Suggestion to streamline hint text, and move formats list to end of message for readability.

ironprogrammer commented:
Thanks for the PR, @madhusudhand! Note that in my testing, I didn't see any error messages, only the success message.

## Test Report

### Environment
- Hardware: MacBook Pro Apple M1 Pro
- OS: macOS 13.5.2
- Browser: Safari 16.6
- Server: nginx/1.25.2
- PHP: 7.4.33
- WordPress: 6.4-alpha-56267-src
- Theme: twentytwentythree v1.2
- Active Plugins:
    - gutenberg v16.7.0-rc.1 (this PR)

### Actual Results
- ✅ Uploads dropzone is located on a new tab (see Figure 1).
- ✅ After successful font upload, "Upload successful" message appears.
- ✅ No snackbars were observed during font upload testing.
- ❌ No error displayed when invalid filetype is drag-n-dropped (tried TXT, ZIP, SVG, EOT).
- ⚠️ No warning feedback when mixed valid/invalid filetypes are drag-n-dropped (e.g. WOFF2 + SVG). Success message appears if *any* files uploaded successfully (see Figure 2).
- ⚠️ Previous successful response message persists after failed upload attempt (see Figure 3).

### Additional Information
- ✅ I was able to install Inclusive Sans without issue (see report [noted here](https://github.com/WordPress/gutenberg/pull/54655#issuecomment-1730991305)).

### Supplemental Artifacts
<details>
<summary>[Click to Expand] Figure 1: New Uploads dropzone location.</summary>

![trac-54655-test](https://github.com/WordPress/gutenberg/assets/824344/0b6349f6-c554-4e80-a2eb-82dd1a7f0d3f)

</details>

<details>
<summary>[Click to Expand] Figure 2: Mixed valid/invalid filetypes register as successful.</summary>

![trac-54655-test-mixed](https://github.com/WordPress/gutenberg/assets/824344/59cac35f-89f6-4ef9-bfa5-582bb1ef72a3)

</details>

<details>
<summary>[Click to Expand] Figure 3: Success message remains following invalid upload attempt.</summary>

![trac-54655-test-persist](https://github.com/WordPress/gutenberg/assets/824344/91b660bd-b5f7-40a8-ae9e-c13be5402035)

</details>

ironprogrammer reviewed (commented):
This is looking nice, @madhusudhand! One suggestion for the hint text here:

madhusudhand committed:
update hint text and clear successful message with a fresh upload

madhusudhand commented:
@ironprogrammer Thanks for the detailed feedback.

> No error displayed when invalid filetype is drag-n-dropped (tried TXT, ZIP, SVG, EOT).

We do not have specific information as to why a font upload failed from API where an upload could fail for multiple reasons. All the error messaging is to be handled in [this followup issue](https://github.com/WordPress/gutenberg/issues/54777) by returning proper error codes from the API.

> No warning feedback when mixed valid/invalid filetypes are drag-n-dropped (e.g. WOFF2 + SVG)

For now you would see only a success message because the error notice was removed. It is possible that we will have successful uploads and errors by choosing mix of files. I have put the details on errors [in this comment](https://github.com/WordPress/gutenberg/issues/54777#issuecomment-1735170658). Please share your thoughts.

> Previous successful response message persists after failed upload attempt (see Figure 3).

It is fixed now.

> Suggestion to streamline hint text, and move formats list to end of message for readability.

It is updated as per the suggestion.

mikachan commented:
As this PR hasn't been approved or merged yet and the latest version of Gutenberg is due to be released today, I'm going to remove the `Backport to Gutenberg RC` label. Please let me know if this doesn't sound right!

I'm going to add the `Backport to WP Beta/RC` label as the Font Library work is a "blessed" task for WordPress 6.4.

matiasbenedetto reviewed (approved):
This is looking good! The problems existing around not displaying an error message when a font upload fails are not in the scope of this PR so let's merge and work on the follow-ups for the endpoints.

mikachan commented:
I just cherry-picked this PR to the commits-for-6.4-beta2 branch to get it included in the next release: 75cb40f965
